src/main/java/org/skriptlang/skript/bukkit/tags/TagType.java

```diff
@@ -0,0 +1,143 @@
+package org.skriptlang.skript.bukkit.tags;
+
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.util.Kleenean;
+import org.bukkit.Keyed;
+import org.bukkit.Material;
+import org.bukkit.entity.EntityType;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.UnmodifiableView;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * The type of a tag. Represents a category or context that the tags apply in.
+ * For example, {@link #ITEMS} tags apply to {@link Material}s, like {@link #BLOCKS}, but in item form rather than as
+ * placed blocks.
+ * <br>
+ * This class also contains a static registry of all tag types.
+ *
+ * @param <T> see type.
+ */
+public class TagType<T extends Keyed> {
+
+	private static final List<TagType<?>> REGISTERED_TAG_TYPES = Collections.synchronizedList(new ArrayList<>());
+
+	public static final TagType<Material> ITEMS = new TagType<>("item", Material.class);
+	public static final TagType<Material> BLOCKS = new TagType<>("block", Material.class);
+	public static final TagType<EntityType> ENTITIES = new TagType<>("entity [type]", "entity type", EntityType.class);
+
+	static {
+		TagType.addType(ITEMS, BLOCKS, ENTITIES);
+	}
+
+	private final String pattern;
+	private final String toString;
+	private final Class<T> type;
+
+	/**
+	 * @param pattern The pattern to use when constructing the selection Skript pattern.
+	 * @param type The class this tag type applies to.
+	 */
+	public TagType(String pattern, Class<T> type) {
+		this(pattern, pattern, type);
+	}
+
+	/**
+	 * @param pattern The pattern to use when constructing the selection Skript pattern.
+	 * @param toString The string to use when printing a toString.
+	 * @param type The class this tag type applies to.
+	 */
+	public TagType(String pattern, String toString, Class<T> type) {
+		this.pattern = pattern;
+		this.type = type;
+		this.toString = toString;
+	}
+
+	public String pattern() {
+		return pattern;
+	}
+
+	public Class<T> type() {
+		return type;
+	}
+
+	@Override
+	public String toString() {
+		return toString;
+	}
+
+	/**
+	 * Adds types to the registered tag types.
+	 * @param type The types to add.
+	 */
+	public static void addType(TagType<?>... type) {
+		synchronized (REGISTERED_TAG_TYPES) {
+			REGISTERED_TAG_TYPES.addAll(List.of(type));
+		}
```

Comment:
```suggestion
		REGISTERED_TAG_TYPES.addAll(List.of(type));
```
Since the list itself is wrapped